src/azul/service/manifest_service.py

```diff
@@ -885,23 +884,32 @@ def write_to(self, output: IO[str]) -> Optional[str]:
             for bundle in list(doc['bundles'])[0:100]:  # iterate over copy …
                 doc['bundles'] = [bundle]  # … to facilitate this in-place modification
                 row = {}
+                related_rows = []
                 for doc_path, column_mapping in self.manifest_config.items():
                     entities = self._get_entities(doc_path, doc)
                     self._extract_fields(entities, column_mapping, row)
+                    if doc_path == 'contents.files':
+                        assert len(entities) == 1
+                        entity = one(entities)
+                        try:
+                            related_files = entity['related_files']
+                        except KeyError:
+                            # Disregard file entities without related_files
+                            continue
+                        else:
+                            for file in related_files:
+                                related_row = {}
+                                for k, v in file.items():
+                                    entity[k] = v if v else entity[k]
```

Comment:
just a question: Do all the related files share the same `content-type`, `file_format`, `file_type` as the entity? 
It looks like `related_files` are only applicable to files in a zarry.

Reply:
if a file is missing a value, and we use the value from the entity, can that create malformed data?

Say if for some reason the `uuid` was `''`, then we would use the `uuid` from the entity, it seems like this can cause malformed data in the manifest. 

You might be able to rework this to compose a dictionary that contains all the stuff from the `files` and cherry pick some stuff out of the entity if it is not present in the file keys.

Reply:
> Do all the related files share the same `content-type`, `file_format`, `file_type` as the entity?

I believe so, by looking at the [object](https://github.com/DataBiosphere/azul/blob/develop/src/azul/plugins/metadata/hca/transform.py#L757-L765) returned by the `_related_file` function, which drives the format of the `related_files` content …
```
{
    'name': file.manifest_entry.name,
    'crc32c': file.manifest_entry.crc32c,
    'sha256': file.manifest_entry.sha256,
    'size': file.manifest_entry.size,
    'uuid': file.manifest_entry.uuid,
    'drs_path': self.bundle.drs_path(file.manifest_entry.json),
    'version': file.manifest_entry.version,
}
```
… and compare it with the object returned by the `_file` [object](https://github.com/DataBiosphere/azul/blob/develop/src/azul/plugins/metadata/hca/transform.py#L718-L744) …
```
{
    'content-type': file.manifest_entry.content_type,
    'indexed': file.manifest_entry.indexed,
    'name': file.manifest_entry.name,
    'crc32c': file.manifest_entry.crc32c,
    'sha256': file.manifest_entry.sha256,
    'size': file.manifest_entry.size,
    'uuid': file.manifest_entry.uuid,
    'drs_path': self.bundle.drs_path(file.manifest_entry.json),
    'version': file.manifest_entry.version,
    'document_id': str(file.document_id),
    'file_type': file.schema_name,
    'file_format': file.file_format,
    'content_description': sorted(file.content_description),
    'source': Submitter.title_for_id(file.submitter_id),
    '_type': 'file',
    'related_files': list(map(self._related_file, related_files)),
    **(
        {
             'read_index': file.read_index,
             'lane_index': file.lane_index
        } if isinstance(file, api.SequenceFile) else {
        }
    ),
}
```
… we can determine that fields not specified in `_related_files` function are copied over to the related file row. It doesn't only use the fields you mention, it also copies all the other entities from the original file, projects, samples etc., as you provably saw in line 913.

>It looks like related_files are only applicable to files in a zarry.

Agreed, however, if I recall correctly, Hannes had previously mentioned that at the moment the `related_files` hierarchy is only exhibited in zarry (as you point out) but in the future other file formats might also have related files.

>if a file is missing a value, and we use the value from the entity, can that create malformed data?

🤔 seems like it, but related files are indexed with a very small subset of fields compared to a full file document in the index. I think that the fields indexed are just enough to locate the file, as such, if the related file had a malformed (missing `uuid` or `sha`) field, something very file metadata specific, Azul would fail to index it.

>…  rework this to compose a dictionary that contains all the stuff from the files and cherry pick some stuff out of the entity if it is not present in the file keys.

If for some reason the the fields that identify the related files were allowed to be absent and the indexer doesn't complain (which can cause the malformation you point out), it might be inevitable to recycle all the other entity fields from the original document. Original and related files together form a single source of data, which is why, I think, they share all the other entities. The other option, which I had accidentally implemented first, was to only include in the row, fields which were populated in the object returned by the `_related_file` function. But that felt like the related files were alienated, since they had less than seven fields populated per row. I am not necessarily sure how to proceed, since this malformations might still occur either way.

Reply:
Some of the fields for the related files are not shared with the entity, one example would be the content-type.
In the example below the entity content type is `'application/json; dcp-type=data'` while one of the related files has `'application/octet-stream; dcp-type=data'`. 

The manifest rows will have mixed fields from the entity and the related files, because https://github.com/DataBiosphere/azul/blob/develop/src/azul/plugins/metadata/hca/transform.py#L757-L765 does not specify all the fields that the file has. 

It might be worthwhile to discuss this with @hannes-ucsc, but I would think that blending the fields like this is not desirable.

I wonder why we even have `_related_files`, and why we cant just use another call to `_files` for this mapping.
https://github.com/DataBiosphere/azul/blob/96105bce814cf18063916020fb48d1f12c733938/src/azul/plugins/metadata/hca/transform.py#L736
 

![Screen Shot 2021-03-11 at 3 15 07 PM](https://user-images.githubusercontent.com/1947534/110869177-c5be3800-827e-11eb-8245-bdeab112d8f6.png)

Sorry about the screenshot size, you might have to use the zoom feature, and get a magnifying glass out too.

Reply:
Include entries that begin with `file.manifest_entry` (included in `_file`) which are currently missing in related files.